File: prettier-plugin-csharpier/src/Printer/Types/JoinClause.ts
import { PrintMethod } from "../PrintMethod";
import { printIdentifier, printPathIdentifier, SyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";

export interface JoinClauseNode extends SyntaxTreeNode<"JoinClause"> {
    joinKeyword?: SyntaxToken;
    type?: SyntaxTreeNode;
    identifier: SyntaxToken;
    inKeyword?: SyntaxToken;
    inExpression?: SyntaxTreeNode;
    onKeyword?: SyntaxToken;
    leftExpression?: SyntaxTreeNode;
    equalsKeyword?: SyntaxToken;
    rightExpression?: SyntaxTreeNode;
    into?: JoinIntoClauseNode;
}

interface JoinIntoClauseNode extends SyntaxTreeNode<"JoinIntoClause"> {
    intoKeyword?: SyntaxToken;
    identifier: SyntaxToken;
}

export const printJoinClause: PrintMethod<JoinClauseNode> = (path, options, print) => {
    const node = path.getValue();
    return concat([
        "join ",
        printIdentifier(node),
        " in ",
        path.call(print, "inExpression"),
        " on ",
        path.call(print, "leftExpression"),
        " equals ",
        path.call(print, "rightExpression"),
        node.into ? " into " + printIdentifier(node.into) : "",
    ]);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/JoinClause.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class JoinClause
{
    public static Doc Print(JoinClauseSyntax node, PrintingContext context)
    {
        return Doc.Concat(
            ExtraNewLines.Print(node),
            Doc.Group(
                Token.PrintWithSuffix(node.JoinKeyword, " ", context),
                node.Type != null ? Doc.Concat(Node.Print(node.Type, context), " ") : Doc.Null,
                Token.PrintWithSuffix(node.Identifier, " ", context),
                Token.PrintWithSuffix(node.InKeyword, " ", context),
                Node.Print(node.InExpression, context),
                Doc.Indent(
                    Doc.Line,
                    Token.PrintWithSuffix(node.OnKeyword, " ", context),
                    Node.Print(node.LeftExpression, context),
                    " ",
                    Token.PrintWithSuffix(node.EqualsKeyword, " ", context),
                    Node.Print(node.RightExpression, context),
                    node.Into != null
                        ? Doc.Concat(
                            Doc.Line,
                            Token.PrintWithSuffix(node.Into.IntoKeyword, " ", context),
                            Token.Print(node.Into.Identifier, context)
                        )
                        : Doc.Null
                )
            )
        );
    }
}
